sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -332,47 +334,60 @@ private[hive] object HadoopTableReader extends HiveInspectors with Logging {
 
     logDebug(soi.toString)
 
+    val allStructFieldNames = soi.getAllStructFieldRefs().toList
+      .map(fieldRef => fieldRef.getFieldName())
+
     val (fieldRefs, fieldOrdinals) = nonPartitionKeyAttrs.map { case (attr, ordinal) =>
-      soi.getStructFieldRef(attr.name) -> ordinal
+      // If the partition contain this attribute or not
```

Comment:
When a non-partition key attribution doesn't exist in the partition's (determined by the partition's `StructObjectInspector`), we should produce a null ref. Previously, we don't check it and directly call `getStructFieldRef`  to query the field ref. It will cause the error reported in the JIRA.